packages/purgecss/src/types/index.ts

```diff
@@ -29,6 +29,96 @@ export interface ExtractorResultDetailed {
   undetermined: string[];
 }
 
+function mergeSets(into: Set<string>, from?: string[] | Set<string>) {
+  if (from) {
+    from.forEach(into.add, into);
+  }
+}
+
+export class ExtractorResultSets {
+  private undetermined = new Set<string>();
+  private attrNames = new Set<string>();
+  private attrValues = new Set<string>();
+  private classes = new Set<string>();
+  private ids = new Set<string>();
+  private tags = new Set<string>();
+
+  constructor(er: ExtractorResult) {
+    this.merge(er);
+  }
+
+  merge(that: ExtractorResult | ExtractorResultSets) {
+    if (Array.isArray(that)) {
+      mergeSets(this.undetermined, that);
+    } else if (that instanceof ExtractorResultSets) {
+      mergeSets(this.undetermined, that.undetermined);
+      mergeSets(this.attrNames, that.attrNames);
+      mergeSets(this.attrValues, that.attrValues);
+      mergeSets(this.classes, that.classes);
+      mergeSets(this.ids, that.ids);
+      mergeSets(this.tags, that.tags);
+    } else {
+      // ExtractorResultDetailed:
+      mergeSets(this.undetermined, that.undetermined);
+      if (that.attributes) {
+        mergeSets(this.attrNames, that.attributes.names);
+        mergeSets(this.attrValues, that.attributes.values);
+      }
+      mergeSets(this.classes, that.classes);
+      mergeSets(this.ids, that.ids);
+      mergeSets(this.tags, that.tags);
+    }
+    return this;
+  }
+
+  hasAttrName(name: string): boolean {
+    return this.attrNames.has(name) || this.undetermined.has(name);
+  }
+
+  private someAttrValue(predicate: (value: string) => boolean): boolean {
+    const forEachFn = (value: string) => {
+      if (predicate(value)) throw true;
+    };
+    try {
+      // The only way to break early from a Set.prototype.forEach
+      // loop is to throw from the callback function.
```

Comment:
Is this better than using a `for...of`?

Reply:
I tend to avoid `for...of` because of how it gets compiled for browsers (lots of code, including nested `try`-`catch`-`finally` blocks!), but this code will be mostly running in Node, which has native support for `for`-`of` loops now. Also, this code is only triggered when there's an attribute selector like `~=` or `^=`, which isn't nearly as common as a simple selector. In other words, I'm happy to switch to `for`-`of` here.

Reply:
oki, `try catch` is fine. Never looked at the compiled `for...of`, thanks for the info. I'll merge this PR now :)